examples/lock-app/qpg6100/src/AppTask.cpp

```diff
@@ -0,0 +1,522 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    Copyright (c) 2019 Google LLC.
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "qvCHIP.h"
+
+#include "AppConfig.h"
+#include "AppEvent.h"
+#include "AppTask.h"
+
+#include "QRCodeUtil.h"
+
+#include "Server.h"
+#include "attribute-storage.h"
+#include "gen/cluster-id.h"
+
+#include <setup_payload/QRCodeSetupPayloadGenerator.h>
+#include <setup_payload/SetupPayload.h>
+
+using namespace chip::TLV;
+using namespace chip::DeviceLayer;
+
+#include <platform/CHIPDeviceLayer.h>
+#if CHIP_ENABLE_OPENTHREAD
+#include <platform/OpenThread/OpenThreadUtils.h>
+#include <platform/ThreadStackManager.h>
+#include <platform/internal/DeviceNetworkInfo.h>
+#include <platform/qpg6100/ThreadStackManagerImpl.h>
+#define JOINER_START_TRIGGER_TIMEOUT 1500
+#endif
+
+#define FACTORY_RESET_TRIGGER_TIMEOUT 3000
+#define FACTORY_RESET_CANCEL_WINDOW_TIMEOUT 3000
+#define APP_TASK_STACK_SIZE (4096)
+#define APP_TASK_PRIORITY 2
+#define APP_EVENT_QUEUE_SIZE 10
+
+TimerHandle_t sFunctionTimer; // FreeRTOS app sw timer.
+
+static TaskHandle_t sAppTaskHandle;
+static QueueHandle_t sAppEventQueue;
+
+static bool sIsThreadProvisioned     = false;
+static bool sIsThreadEnabled         = false;
+static bool sIsThreadAttached        = false;
+static bool sIsPairedToAccount       = false;
+static bool sHaveBLEConnections      = false;
+static bool sHaveServiceConnectivity = false;
+
+AppTask AppTask::sAppTask;
+
+int AppTask::StartAppTask()
+{
+    int err = CHIP_ERROR_MAX;
+
+    sAppEventQueue = xQueueCreate(APP_EVENT_QUEUE_SIZE, sizeof(AppEvent));
+    if (sAppEventQueue == NULL)
+    {
+        ChipLogError(NotSpecified, "Failed to allocate app event queue");
+        // appError(err);
+    }
+
+    // Start App task.
+    if (xTaskCreate(AppTaskMain, "APP", APP_TASK_STACK_SIZE / sizeof(StackType_t), NULL, 1, &sAppTaskHandle) == pdPASS)
+    {
+        err = CHIP_NO_ERROR;
+    }
+
+    return err;
+}
+
+int AppTask::Init()
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    // Subscribe with our button callback to the qvCHIP button handler.
+    qvCHIP_SetBtnCallback(ButtonEventHandler);
+
+    // Create FreeRTOS sw timer for Function Selection.
+    sFunctionTimer = xTimerCreate("FnTmr",          // Just a text name, not used by the RTOS kernel
```

Comment:
My understanding is that we are generally preferring CHIP timers to raw FreeRTOS timers.... and yes, the current example apps (except esp32) don't do a great job of that.

Reply:
You are referring to the SystemLayer - StartTimer()/CancelTimer() functions I assume.
Indeed used in ESP32. I'll check to replace the functionality here.
Prep for #3840 as well

Reply:
Updated to use the SystemLayer timer. 
For the FreeRTOS task and queue I don't see an immediate substitution available. 
There is a Platform PostEvent() that could subsitute the queue, but is that intended to be used at application level ?

Reply:
I don't know offhand, sorry....